### PR TITLE
Fix url bugs

### PR DESCRIPTION
**Pull request by mruwnik**
https://github.com/StampyAI/stampy-ui/issues/161
https://github.com/StampyAI/stampy-ui/issues/160

### CODE REVIEW

File: app/hooks/stateModifiers.tsx [hunk 1]
```diff
@@ -5,11 +5,25 @@ type StateString = string
...
-  ])
+  Array.from(state.matchAll(/([^-_r]+)([-_r]*)/g) ?? [])
+    .map(
+      (groups) =>
+        [
+          groups[1], // question id
+          (groups[2] as QuestionState) || QuestionState.OPEN,
+        ] as StateEntry
+    )
+    .filter(uniqueChecker())
```
Comment (mruwnik) [inline app/hooks/stateModifiers.tsx:21]: make sure that duplicates get removed from the url string. The first one found is the value (and position) in the resulting url string

File: app/hooks/stateModifiers.tsx [hunk 2]
```diff
@@ -5,11 +5,25 @@ type StateString = string
 
 export const TOP = '-1'
 
+const uniqueChecker = (): ((e: StateEntry) => boolean) => {
```
Comment (Aprillion) [inline app/hooks/stateModifiers.tsx:8]: can we call it `makeUniqueChecker` or `uniqueCheckerFactory`?

File: app/hooks/stateModifiers.tsx [hunk 3]
```diff
@@ -5,11 +5,25 @@ type StateString = string
...
+          groups[1], // question id
+          (groups[2] as QuestionState) || QuestionState.OPEN,
+        ] as StateEntry
+    )
+    .filter(uniqueChecker())
+
+export const processStateEntries = (
+  state: StateString,
+  func = (e: StateEntry[]): StateEntry[] => e
+): StateString => func(getStateEntries(state)).filter(uniqueChecker()).flat().join('')
```
Comment (Aprillion) [inline app/hooks/stateModifiers.tsx:26]: does it need double filtering, both inside `getStateEntries` and after?

**Review by Aprillion — APPROVED** [review]

File: app/hooks/stateModifiers.tsx (hunk 3 above)
Comment (mruwnik) [reply app/hooks/stateModifiers.tsx]: yes and no :/  which means that can be tidied up

File: app/hooks/stateModifiers.tsx (hunk 2 above)
Comment (mruwnik) [reply app/hooks/stateModifiers.tsx]: changed

File: app/hooks/useQuestionStateInUrl.ts [hunk 4]
```diff
@@ -28,7 +29,9 @@ export default function useQuestionStateInUrl(minLogo: boolean, initialQuestions
   const [remixSearchParams] = useSearchParams()
   const transition = useTransition()
 
-  const [stateString, setStateString] = useState(() => remixSearchParams.get('state'))
+  const [stateString, setStateString] = useState(
+    () => remixSearchParams.get('state') && processStateEntries(remixSearchParams.get('state'))
```
[the remark this answers is not in the document]
Comment (mruwnik) [reply app/hooks/useQuestionStateInUrl.ts:33]: it would, but would require changing the type, which would then mess up the other places...

**Review by Aprillion — APPROVED** [review]

**Pull request merged**